lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1680,8 +1723,18 @@ def loki_endpoints(self) -> List[dict]:
             A list with Loki Push API endpoints.
         """
         endpoints = []  # type: list
+
         for relation in self._charm.model.relations[self._relation_name]:
-            endpoints = endpoints + json.loads(relation.data[relation.app].get("endpoints", "[]"))
+            for unit in relation.units:
+                if unit.app is self._charm.app:
+                    # This is a peer unit
+                    continue
+
+                endpoint = relation.data[unit].get("endpoint")
+                if endpoint:
+                    deserialized_endpoint = json.loads(endpoint)
+                    endpoints.append(deserialized_endpoint)
```

Comment:
Would this produce the desired behavior in the case of several related loki _apps_?

Reply:
Yes.

An `agent` (grafana agent) related to 4 different loki apps:

```
Model  Controller  Cloud/Region        Version  SLA          Timestamp
cos    charm-dev   microk8s/localhost  2.9.25   unsupported  17:54:14-03:00

App     Version  Status  Scale  Charm              Channel  Rev  Address         Exposed  Message
agent            active      1  grafana-agent-k8s             2  10.152.183.51   no
jose             active      1  loki-k8s                     14  10.152.183.135  no
kilolo           active      2  loki-k8s                     12  10.152.183.58   no
leon             active      1  loki-k8s                     13  10.152.183.117  no
loki             active      2  loki-k8s                     11  10.152.183.164  no

Unit       Workload  Agent  Address       Ports  Message
agent/0*   active    idle   10.1.157.76
jose/0*    active    idle   10.1.157.100
kilolo/0   active    idle   10.1.157.101
kilolo/1*  active    idle   10.1.157.85
leon/0*    active    idle   10.1.157.102
loki/0*    active    idle   10.1.157.119
loki/1     active    idle   10.1.157.65

Relation provider  Requirer                Interface      Type     Message
jose:logging       agent:logging-consumer  loki_push_api  regular
kilolo:logging     agent:logging-consumer  loki_push_api  regular
leon:logging       agent:logging-consumer  loki_push_api  regular
loki:logging       agent:logging-consumer  loki_push_api  regular
```

The config produced in in agent config file:

```yaml
loki:
  configs:
  - clients:
    - url: http://loki-0.loki-endpoints.cos.svc.cluster.local:3100/loki/api/v1/push
    - url: http://loki-1.loki-endpoints.cos.svc.cluster.local:3100/loki/api/v1/push
    - url: http://kilolo-1.kilolo-endpoints.cos.svc.cluster.local:3100/loki/api/v1/push
    - url: http://kilolo-0.kilolo-endpoints.cos.svc.cluster.local:3100/loki/api/v1/push
    - url: http://leon-0.leon-endpoints.cos.svc.cluster.local:3100/loki/api/v1/push
    - url: http://jose-0.jose-endpoints.cos.svc.cluster.local:3100/loki/api/v1/push
    name: promtail
    positions:
      filename: /tmp/positions.yaml
    scrape_configs:
    - job_name: loki
      loki_push_api:
        server:
          grpc_listen_port: 3600
          http_listen_port: 3500
```

Reply:
Nice!
Does it make sense for `loki_endpoints` to flatten all endpoints? I.e. is there a reason to aggregate them by relation, and have the charm code flatten it?

Reply:
I think I'm not getting your point... what do you mean with: "flatten all endpoints" ??

Reply:
I am not too familiar with the agent config but I was wondering if it makes sense for the lib to return
```python
[{"url": "http://..."}, ...]
```
per relation. For example:
```python
[
  [{"url": "url: http://loki-0"}, {"url": "url: http://loki-0"}],
  [{"url": "url: http://kilolo-0"}, {"url": "url: http://kilolo-0"}],
]
```
Not sure if it is needed though or if backwards compat is a consideration here.

Reply:
I think it is not necessary, since the `loki_endpoints` method in `LokiPushApiConsumer` returns endpoints for the current relation